store/consul/consul.go

```diff
@@ -379,44 +382,95 @@ func (s *Consul) NewLock(key string, options *store.LockOptions) (store.Locker,
-				TTL:       (options.TTL / 2).String(), // Consul multiplies the TTL by 2x
-				LockDelay: 1 * time.Millisecond,       // Virtually disable lock delay
-			}
-
-			// Create the key session
-			session, _, err := s.client.Session().Create(entry, nil)
-			if err != nil {
-				return nil, err
-			}
-
-			// Place the session on lock
-			lockOpts.Session = session
-
-			// Renew the session ttl lock periodically
-			go s.client.Session().RenewPeriodic(entry.TTL, session, nil, options.RenewLock)
-			lock.renewCh = options.RenewLock
+			ttl = options.TTL
 		}
-
 		// Set optional value on Lock
 		if options.Value != nil {
 			lockOpts.Value = options.Value
 		}
 	}
 
+	entry := &api.SessionEntry{
+		Behavior:  api.SessionBehaviorRelease, // Release the lock when the session expires
+		TTL:       (ttl / 2).String(),         // Consul multiplies the TTL by 2x
+		LockDelay: 1 * time.Millisecond,       // Virtually disable lock delay
+	}
+
+	// Create the key session
+	session, _, err := s.client.Session().Create(entry, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Place the session and renew chan on lock
+	lockOpts.Session = session
+	lock.renewCh = options.RenewLock
+
 	l, err := s.client.LockOpts(lockOpts)
 	if err != nil {
 		return nil, err
 	}
 
+	// Renew the session ttl lock periodically
+	s.renewLockSession(entry.TTL, session, options.RenewLock)
+
 	lock.lock = l
 	return lock, nil
 }
 
+// renewLockSession is used to renew a session Lock, it takes
+// a stopRenew chan which is used to explicitely stop the session
+// renew process. The renew routine never stops until a signal is
+// sent to this channel. If deleting the session fails because the
+// connection to the store is lost, it keeps trying to delete the
+// session periodically until it can contact the store, this ensures
+// that the lock is not maintained indefinitely which ensures liveness
+// over safety for the lock when the store becomes unavailable. Keep
+// in mind that if the connection to the store is lost forever, you
+// will leak the routine trying to renew the lock session.
+func (s *Consul) renewLockSession(initialTTL string, id string, stopRenew chan struct{}) {
+	ttl, err := time.ParseDuration(initialTTL)
+	if err != nil {
+		return
+	}
+	go func() {
+		for {
+			select {
+			case <-time.After(ttl / 2):
+				entry, _, err := s.client.Session().Renew(id, nil)
+				if err != nil {
+					// If an error occurs, continue until the
+					// session gets destroyed explicitely or
+					// the session ttl times out
+					continue
+				}
+				if entry == nil {
+					return
+				}
+
+				// Handle the server updating the TTL
+				ttl, _ = time.ParseDuration(entry.TTL)
+
+			case <-stopRenew:
+				// Attempt a session destroy
+				_, err := s.client.Session().Destroy(id, nil)
+				if err == nil {
+					return
+				}
+
+				// We can't destroy the session because the store
+				// is unavailable, wait for the session renew period
+				time.Sleep(ttl / 2)
```

Comment:
Took care of the comment

The goroutine is cleaned up after a maximum number of attempts. Only downside is a slower recovery on failure because you might have to wait for the entire TTL expiration period to recover a lock after failing to destroy the session explicitely, but sounds more reasonable than leaking goroutines after losing connectivity to the store that might never be recovered.